src/blocks/battery.rs

```diff
@@ -191,13 +206,154 @@ impl PowerSupplyDevice {
+
+        // https://upower.freedesktop.org/docs/Device.html#Device:Type
+        if upower_type.0 != 2 {
+            return Err(BlockError(
+                "battery".into(),
+                "UPower device is not a battery.".into(),
+            ));
+        }
+        Ok(UpowerDevice {
+            device_path: device_path,
+            con: con,
+        })
+    }
+
+    /// Monitor UPower property changes in a separate thread and send updates
+    /// via the `update_request` channel.
+    pub fn monitor(&self, id: String, update_request: Sender<Task>) {
+        let path = self.device_path.clone();
+        thread::spawn(move || {
+            let con = dbus::Connection::get_private(dbus::BusType::System)
+                .expect("Failed to establish D-Bus connection.");
+            let rule = format!(
+                "type='signal',\
+                 path='{}',\
+                 interface='org.freedesktop.DBus.Properties',\
+                 member='PropertiesChanged'",
+                path
+            );
+
+            // First we're going to get an (irrelevant) NameAcquired event.
+            con.incoming(10_000).next();
+
+            con.add_match(&rule)
+                .expect("Failed to add D-Bus match rule.");
+
+            loop {
+                if let Some(_) = con.incoming(10_000).next() {
+                    update_request.send(Task {
+                        id: id.clone(),
+                        update_time: Instant::now(),
+                    });
+                    // Avoid update spam.
+                    // TODO: Is this necessary?
+                    thread::sleep(Duration::from_millis(1000))
+                }
+            }
+        });
+    }
+}
+
+impl BatteryDevice for UpowerDevice {
+    fn status(&self) -> Result<String> {
+        let status: dbus::arg::Variant<u32> =
+            get_upower_property(&self.con, &self.device_path, "State")?
+                .get1()
+                .block_error("battery", "Failed to read UPower State property.")?;
+
+        // https://upower.freedesktop.org/docs/Device.html#Device:State
+        match status.0 {
+            1 => Ok("Charging".to_string()),
+            2 => Ok("Discharging".to_string()),
+            3 => Ok("Empty".to_string()),
+            4 => Ok("Full".to_string()),
+            5 => Ok("Not charging".to_string()),
+            6 => Ok("Discharging".to_string()),
+            _ => Ok("Unknown".to_string()),
+        }
+    }
+
+    fn capacity(&self) -> Result<u64> {
+        let capacity: dbus::arg::Variant<f64> =
+            get_upower_property(&self.con, &self.device_path, "Percentage")?
+                .get1()
+                .block_error("battery", "Failed to read UPower Percentage property.")?;
+
+        if capacity.0 > 100.0 {
+            Ok(100)
+        } else {
+            Ok(capacity.0 as u64)
+        }
+    }
+
+    fn time_remaining(&self) -> Result<u64> {
```

Comment:
Hi @atheriel i just noticed that "Time to full" does not work with this PR even though upower does support it.

This looks like the "Remaining time" (the one to be shown) is only set to the "TimeToEmpty" here.

It should be set to "TimeToFull" while charging.

Reply:
Ah, you are correct.